RavenDB-18353 Alphanumerical sorting - server side

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18353

### Additional description

Integration of comparer.

### Type of change

- New feature


### How risky is the change?


- Not relevant

### Backward compatibility


- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works


### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

redknightlois committed:
RavenDB-18353: Implementation of the Alphanumeric Natural Comparer

maciejaszyk committed:
RavenDB-18353 Integration of alphanumerical sorting.

ayende commented on src/Corax/Queries/IMatchComparer.cs:
```diff
@@ -3,12 +3,19 @@
 
 namespace Corax.Queries
 {
+    public enum SortingType
```
why is this needed?

maciejaszyk replied:
This is a workaround for the current Corax API.

@redknightlois can you add `Alphanumeric` to `MatchCompareFieldType`? Because alphanumeric will always operate on strings, I think.

redknightlois replied:
Natural Order is only reasonable to work when you are doing roman language encodings (ASCII). I dont think we need to include anything else, this is kinda a replacement of the normal sorting only on certain occasions (should be Opt-in) but that is something that only RavenDB can know.

ayende commented on src/Corax/Queries/SortingMatch.Comparers.cs:
```diff
@@ -20,7 +15,158 @@ internal static int CompareAscending(ReadOnlySpan<byte> x, ReadOnlySpan<byte> y)
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static int CompareAlphanumericAscending(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b, bool foldCase = false)
```
This is valid only to ASCII buffers

maciejaszyk replied:
@redknightlois

redknightlois replied:
Yes, we need to check the buffer for ASCIIness and go back to standard comparer. Unless you know how to do that with Unicode (UTF-8).

ayende replied:
Alpha numeric is rare enough that we don't need to specialize here. Can always use the usual comparere.

redknightlois committed:
RavenDB-18353: Support for UTF-8 strings.

We will switch into Unicode comparison (chars) when we find out that we are not dealing with a ASCII string.

redknightlois commented:
Support for UTF8: https://github.com/redknightlois/ravendb/commit/16afc7c6d6325f73c2655d063fd95886a6831e99
cc @maciejaszyk

maciejaszyk committed:
RavenDB-18353 Tests

redknightlois committed:
RavenDB-18353: Very subtle difference to deal with capital letters.

redknightlois commented:
Commit for fixed handling of capital letters: 6ed9f52c8d03e4d80364afcf2c39fe4fbad35eba
Still needed to update the tests to expect differences with Lucene.

maciejaszyk committed:
RavenDB-18353 SortingMatch cleanup + alphanumeric test
